File: src/main/golang/leetcode/editor/cn/[304]Range Sum Query 2D - Immutable.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
type NumMatrix struct {
	prefix [][]int
}

func Constructor(matrix [][]int) NumMatrix {
	m, n := len(matrix), len(matrix[0])
	p := make([][]int, m+1)
	p[0] = make([]int, n+1)
	for i := 1; i <= m; i++ {
		p[i] = make([]int, n+1)
		for j := 1; j <= n; j++ {
			p[i][j] = p[i-1][j] + p[i][j-1] + matrix[i-1][j-1] - p[i-1][j-1]
		}
	}
	return NumMatrix{prefix: p}
}

func (this *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
	return this.prefix[row2+1][col2+1] - this.prefix[row1][col2+1] - this.prefix[row2+1][col1] + this.prefix[row1][col1]
}

/**
 * Your NumMatrix object will be instantiated and called as such:
 * obj := Constructor(matrix);
 * param_1 := obj.SumRegion(row1,col1,row2,col2);
 */
//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[304]Range Sum Query 2D - Immutable.py
from typing import List


# leetcode submit region begin(Prohibit modification and deletion)
class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        self.sums = [[0] * (len(matrix[0]) + 1) for _ in range(len(matrix) + 1)]
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                self.sums[i + 1][j + 1] = self.sums[i + 1][j] + self.sums[i][j + 1] + matrix[i][j] - self.sums[i][j]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        return (self.sums[row2 + 1][col2 + 1] - self.sums[row2 + 1][col1] - self.sums[row1][col2 + 1] +
                self.sums[row1][col1])

# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
# leetcode submit region end(Prohibit modification and deletion)
